[HLSL][RootSignature] Implement initial parsing of the descriptor table clause params

inbelic opened the pull request:
- Defines `ParseDescriptorTableClauseParams` to establish the pattern of how we will parse parameters in root signatures. Namely, to use recursive descent parsing in a way that follows closely to the EBNF notation definition in the root signature spec.

- Implements parsing of two param types: `UInt32` and `Register` to demonstrate the parsing implementation and allow for unit testing

- Changes the calling convention to use `std::optional` return values instead of boolean error returns and parameters by reference

Part two of implementing: https://github.com/llvm/llvm-project/issues/126569

Commit:
define ParseUInt and ParseRegister to plug into parameters

Commit:
[HLSL][RootSignature] Add infastructure to parse parameters

- defines `ParamType` as a way to represent a reference to some
parameter in a root signature

- defines `ParseParam` and `ParseParams` as an infastructure to define
how the parameters of a given struct should be parsed in an orderless
manner

Commit:
self-review: use consumeExpectedToken api to report unexpected_end_of_params diag

inbelic commented on clang/lib/Parse/ParseHLSLRootSignature.cpp:
```diff
@@ -87,37 +86,174 @@ bool RootSignatureParser::parseDescriptorTableClause() {
+    Register->ViewType = RegisterType::SReg;
+    break;
+  default:
+    break; // Unreachable given Try + assert pattern
```
Whoops, pattern to use `unreachable` didn't get propagated here it seems

Commit:
self-review: clang-format fix up

Commit:
self-review: fix-up comments

Commit:
rebase changes: update to build issue fixes

Commit:
clang-formatting

joaosaffran commented on clang/lib/Parse/ParseHLSLRootSignature.cpp:
```diff
@@ -89,37 +88,178 @@ bool RootSignatureParser::parseDescriptorTableClause() {
+}
+
+bool RootSignatureParser::parseUIntParam(uint32_t *X) {
+  assert(CurToken.TokKind == TokenKind::pu_equal &&
```
This made me confused. The only place this is called is parsing `TokenKind::pu_equal` before calling it, so it seems confusing, IMHO, that this assert is here.

inbelic replied:
The assert is there to communicate that it should only be called after consuming the preceding `=` symbol, when specifying a parameter value `space = u32`, `numDescriptors = u32`.

Maybe you are confused why it is restricted to that case? If so, we can modify the assert to include other valid tokens that come before this token, but keep a clear communication about what is allowed.

joaosaffran commented on clang/lib/Parse/ParseHLSLRootSignature.cpp:
```diff
@@ -89,37 +88,178 @@ bool RootSignatureParser::parseDescriptorTableClause() {
+  return false;
+}
+
+bool RootSignatureParser::handleUIntLiteral(uint32_t *X) {
```
nit: What is the difference between `parseUIntParam` and `parseUIntParam`?

inbelic replied:
`handleUIntLiteral` functions on the `CurToken` to turns its `NumSpelling` field into a `uint32_t`. So for example when parsing the registers here, we also invoke this method on the register token. It is just extracted out for common functionality.

joaosaffran reviewed (commented):
Few questions to help me understand better the changes being introduced here.

llvm-beanz commented on clang/lib/Parse/ParseHLSLRootSignature.cpp:
```diff
@@ -89,37 +88,178 @@ bool RootSignatureParser::parseDescriptorTableClause() {
 
-  if (consumeExpectedToken(TokenKind::pu_r_paren, diag::err_expected_after,
-                           CurToken.TokKind))
+  llvm::SmallDenseMap<TokenKind, ParamType> Params = {
```
Hypothetically, based on the grammar what is the maximum number of token->param mappings we could have here?

The reason I'm asking: is it better to have a map with dynamic allocations and lookup, or just a struct full of optionals to serve as the state?

inbelic replied:
For descriptor table clauses specifically it will grow to 6 params, which would be maintainable as a struct. But if we want to reuse the `parseParams` logic for `StaticSampler` then a common state struct would be around ~20 members.

Notably, this mapping also serves as a mapping to which parse method should be invoked based on the encountered token, it is not immediately clear how we would retain that info using the optionals struct. Although maybe you are implicitly suggesting that we don't have such a generic `parseParams` method.

If we are concerned about dynamic allocations/lookup, the size of this mapping is known statically, so we could also do something like:
```
template <usize N>
struct ParamMap {
  TokenKind Kinds[N];
  ParamType Types[N];
  bool Mandatory[N];
  bool Seen[N];
}
```
WDYT?

llvm-beanz replied:
I'm a bit torn. The pattern used in Clang for things that can be parsed in arbitrary order (like attributes) might be a bit overkill. Clang fully separates parsing from semantic analysis, so it parses all your attributes separate from converting them into AST nodes and attaching them to the proper parent. 

The structure-of-array approach storing all tokens has some scalability concerns to me, and I'm not sure the genericness-gives benefit. Similarly I'm not sure the map approach as implemented is the right approach.

My gut feeling is that we probably shouldn't use the same implementation for parsing descriptor table parameters and sampler parameters.

I think a parameters structure for 6 parameters with either bools (which could be uint32_t bitfield members) or optionals to signal if they are seen makes sense. I suspect if they're mandatory should be something we know from context and don't need additional tracking of.

inbelic replied:
Hm, okay. I think that is where we disagree then: whether it is worthwhile to make this generic or not.

I will make the (opinionated) argument for having it implemented generically:
- There isn't much of a difference in how we will parse the different root parameter types, `RootSBV(...), CBV(...), StaticSampler(...), RootConstants(...), etc`, so, adding their respective parse methods in future prs will just be something like:
```
bool RootSignatureParser::parseStaticSampler() {
  using StaticSamplerParams = ParamMap<13>; // N = # of params locally
  StaticSampler Sampler;
  StaticSamplerParams Params = {
    /* Kinds=*/ = {
      TokenKind::sReg,
      TokenKind::kw_filter,
      ...
    },
    /* ParamType=*/ = {
      &Sampler.Register
      &Sampler.Filter,
      ...
    },
    /*Seen=*/ 0x0,
    /*Mandatory=*/0x1 // only register is required
  };
  if (parseParams(Params))
    return true;
  ...
}
```
- We can contrast that with DXC, [here](https://github.com/microsoft/DirectXShaderCompiler/blob/c940161bb3398ff988fafc343ed1623d4a3fad6c/tools/clang/lib/Parse/HLSLRootSignature.cpp#L1301), where it needs to redefine the same "seen" and "mandatory" functionality over and over again.
- I assume that if we don't want to have a generic method in clang then the code flow would follow a similar pattern as DXC (maybe I don't understand the struct approach correctly and that is a wrong assumption?).
- Therefore, using a generic way to parse the parameters of root parameter types will be clearer in its intent (it is declarative) and easier to extend (instead of copying functionality over) when we add the other parts of the RootSignatureParser

These reasons are why I went with the current generic implementation.

Regarding scalability, the struct of arrays or the map will have a statically known `N` elements (or pairs), where `N` is the number of parameters for a given root parameter type. (`N` is not equal to the total number of token kinds). The largest `N` would be is for `StaticSamplers` with 13, and then for example `DescriptorTableClause` it is 4. And we could make `Mandatory/Seen` just be two `uint32_t` bitfields.

llvm-beanz replied:
> Hm, okay. I think that is where we disagree then: whether it is worthwhile to make this generic or not.
> 
> I will make the (opinionated) argument for having it implemented generically:
> 
> * There isn't much of a difference in how we will parse the different root parameter types, `RootSBV(...), CBV(...), StaticSampler(...), RootConstants(...), etc`, so, adding their respective parse methods in future prs will just be something like:

I don't think what you've described int he code example here is where the genericness becomes a problem.

>  * We can contrast that with DXC, [here](https://github.com/microsoft/DirectXShaderCompiler/blob/c940161bb3398ff988fafc343ed1623d4a3fad6c/tools/clang/lib/Parse/HLSLRootSignature.cpp#L1301), where it needs to redefine the same "seen" and "mandatory" functionality over and over again.
> * I assume that if we don't want to have a generic method in clang then the code flow would follow a similar pattern as DXC (maybe I don't understand the struct approach correctly and that is a wrong assumption?).

I think you draw a false equivalence. You seem to be saying we either do this generically, or we do this like DXC... That's not what I'm saying. The approach that Clang takes generally in parsing is to parse things out, and from the parsed information construct a declaration or statement, which then gets validated during construction.

That is not how DXC's parser for root signatures works, nor is it what you're proposing, but maybe it's the better approach.

> * Therefore, using a generic way to parse the parameters of root parameter types will be clearer in its intent (it is declarative) and easier to extend (instead of copying functionality over) when we add the other parts of the RootSignatureParser

I think the bigger question is at what layer of abstraction does being generic help and at what layer of abstraction is it a hinderance.

Having a `parseRootParam` function that "generically" parses a root-level parameter seems like a good idea, but should it remain generic based on the type of the parameter? Should we have a single "parseArgument" function? Maybe... Should these functions produce the same structural result for every parameter? These things are less clear to me.

> These reasons are why I went with the current generic implementation.
> 
> Regarding scalability, the struct of arrays or the map will have a statically known `N` elements (or pairs), where `N` is the number of parameters for a given root parameter type. (`N` is not equal to the total number of token kinds). The largest `N` would be is for `StaticSamplers` with 13, and then for example `DescriptorTableClause` it is 5. And we could make `Mandatory/Seen` just be two `uint32_t` bitfields.

The reasons that led you to a generic solution, might be reasons why following existing patterns that Clang uses for order-independent parsing (like attributes), might be the better architectural decision. Maybe we should borrow more from Clang's AST design and have polymorphic returned objects rather than variants.

inbelic replied:
> I don't think what you've described int he code example here is where the genericness becomes a problem.

> I think you draw a false equivalence. You seem to be saying we either do this generically, or we do this like DXC... That's not what I'm saying. The approach that Clang takes generally in parsing is to parse things out, and from the parsed information construct a declaration or statement, which then gets validated during construction.
> 
> That is not how DXC's parser for root signatures works, nor is it what you're proposing, but maybe it's the better approach.

I see, then I did misinterpret what the other approach could be.

IIUC in this context, we might have an intermediate object `ParsedParamState` used to represent parameter values. `parseParams` would go through and parse the values to this object. Then we can construct our in-memory structs (`DescriptorTableClause`) from this object and validate the parameter values were specified correctly.

The most recent commit is a prototype of this, expect we allow some validation to be done when parsing.

> I think the bigger question is at what layer of abstraction does being generic help and at what layer of abstraction is it a hinderance.

I think a good case to illustrate what level of abstraction we want is to consider parsing the `flags = FLAGS` parameter, where `FLAGS` could be a variety of flag types (`RootFlags`, `DescriptorRangeFlags`, etc.

Given the root element, (`RootCBV`, `CBV`, etc), it will directly imply the expected flag type.

Imo, we should validate it is the expected flag type during parsing instead of validation. Otherwise, we are throwing out that information and are forced to parse the flags in a general manner, only to rederive what the valid type is later.

Similar with register types, if we have `CBV` we should only parse a `b` register.

This allows for better localized diag messages by raising an error earlier at the invalid register or flag type.

So, I don't think the abstraction to have _all_ validation after parsing is beneficial. But we can have the ones that make sense to be when constructing the elements (checking which are mandatory), as this is more clear.

> Having a `parseRootParam` function that "generically" parses a root-level parameter seems like a good idea, but should it remain generic based on the type of the parameter? Should we have a single "parseArgument" function? Maybe... Should these functions produce the same structural result for every parameter? These things are less clear to me.

> The reasons that led you to a generic solution, might be reasons why following existing patterns that Clang uses for order-independent parsing (like attributes), might be the better architectural decision. Maybe we should borrow more from Clang's AST design and have polymorphic returned objects rather than variants.

I do think that using variants and mapping the variant types to a `parseMethod` is more confusing than it needs to be. Using a stateful struct is more clear in assigning them and easy to follow. And it removes the need for having to work around having an `any` type with polymorphic objects or variants.

joaosaffran reviewed (approved):
LGTM, not an expert, though.

Commit:
review: prototype using a stateful struct for parsed params

pros:
- more explicit mapping to what parse method should be called based on
the current keyword/root element
- removes complexity of using `std::visit` on the parameter types
- allows for validations before the in-memory root element is
constructed

cons:
- we will need to bind the parsed values to the in-memory
representations for each element as opposed to just setting it directly.
but it makes the validations performed more explicit

notes:
- this does not enforce that we should do all validations after parsing
out, we do allow validations on the current token types (for instance
with register types)

inbelic commented on llvm/include/llvm/Frontend/HLSL/HLSLRootSignature.h:
```diff
@@ -32,11 +39,19 @@ struct DescriptorTable {
+// RootElement. Each variant of ParamType is expected to have a Parse method
+// defined that will be dispatched on when we are attempting to parse a
+// parameter
+using ParamType = std::variant<uint32_t *, Register *>;
```
Can be removed completely now if we use the struct approach

bogner commented:
I'm not entirely convinced the generic `ParsedParamState` is the right level of abstraction. In some ways the declarative nature is nice, but I think it's quite a bit more complex than a simple recursive descent here. 

Warning - lots of untested and never compiled code follows. Consider an API like so:

```c++
  struct ParsedParam {
    std::optional<llvm::hlsl::rootsig::Register> Register;
    std::optional<uint32_t> Space;
  };
  std::optional<ParsedParam> parseParameter(TokenKind RegType);
```

This could then be used in `parseDescriptorClause for the various parameter kinds:
```c++
  DescriptorTableClause Clause;
  std::optional<ParsedParam> P;
  switch (ParamKind) {
  default:
    llvm_unreachable("Switch for consumed token was not provided");
  case TokenKind::kw_CBV:
    Clause.Type = ClauseType::CBuffer;
    P = parseParameter(TokenKind::bReg);
    break;
  // ...
  }
  if (!P.has_value())
    // diagnose
  // otherwise we have a parameter!
```

then parseParameter can just can follow the same pattern and just recurse into the various members:
```c++
std::optional<ParsedParam> parseParameter(TokenKind RegType) {
  if (consumeExpectedToken(TokenKind::pu_l_paren, diag::err_expected_after,
                           ParamKind))
    return std::nullopt;

  ParsedParam Result;

  do {
    if (tryConsumeExpectedToken(RegType)) {
      if (Result.Register.has_value()) {
        getDiags().Report(CurToken.TokLoc, diag::err_hlsl_rootsig_repeat_param)
            << CurToken.TokKind;
        return true;
      }
      if (auto Register = parseRegister(Params.Register))
        Result.Register = *Register;
    }
    if (tryConsumeExpectedToken(RootSignatureToken::Kind::kw_space)) {
      if (Result.Space.has_value()) {
        getDiags().Report(CurToken.TokLoc, diag::err_hlsl_rootsig_repeat_param)
            << CurToken.TokKind;
        return true;
      }
      if (auto Space = parseUIntParam())
        Result.Register = *Register;
    }

  } while (tryConsumeExpectedToken(TokenKind::pu_comma));

  if (!consumeExpectedToken(TokenKind::pu_r_paren,
                            diag::err_hlsl_unexpected_end_of_params,
                            /*param of=*/ParamKind))
    return std::nullopt;

  if (!Result.Register.has_value()) {
    getDiags().Report(CurToken.TokLoc, diag::err_hlsl_rootsig_missing_param)
        << ExpectedRegister;
    return std::nullopt;
  }

  return Result;
```

This also makes the functions match the shape of the grammar formulations in the EBNF notation. I suspect that any code duplication we find by doing it this way would be fairly easy to clean up with a few helper functions here and there, and since the functions should stay relatively short I think it keeps it reasonably simple.

Commit:
review: remove abstraction of parsing arbitrary params

Commit:
NFC review: update calling conventions to use std::optional instead of bool

inbelic commented:
I don't have any strong arguments against structuring it like described and have updated accordingly.

Since we are changing the calling convention of the `parse.*`  methods to use `std::optional` I had originally also updated the `parseDescriptorTable` and `parseDescriptorTableClause` as well. But it made the diff difficult to read.

Changes to make the calling convention consistent with those methods are in the branch [here](https://github.com/inbelic/llvm-project/tree/inbelic/rs-update-call-convs). I will create the clean up pr once this one merges
